Unable to build on macOS m2

A fresh install of `pup` fails to build on Apple M2 Sonoma 14.5.

```
(pup) 18:33:31 ✔ cargo clean && cargo build
     Removed 3254 files, 1.6GiB total
    Updating crates.io index
    Updating git repository `https://github.com/huggingface/candle.git`
    Updating git repository `https://github.com/rcastill/mat2image.git`
error: failed to select a version for the requirement `candle-core = "^0.4.0"`
candidate versions found which didn't match: 0.6.0
location searched: Git repository https://github.com/huggingface/candle.git
required by package `pup v0.1.0 (/Users/tallam/github/tallamjr/origin/pup)`

```
After updating the `Cargo.toml` file to match `v0.6.0` for `candle`, the following error occurs:
```
  === Generated: core in 18.196518083s
  thread 'main' panicked at /Users/tallam/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencv-0.88.8/build/generator.rs:119:36:
  Generator process panicked: Any { .. }
  stack backtrace:
     0:        0x1025c421c - std::backtrace_rs::backtrace::libunwind::trace::he4f0a5f56afe8e37
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/../../backtrace/src/backtrace/libunwind.rs:105:5
     1:        0x1025c421c - std::backtrace_rs::backtrace::trace_unsynchronized::habb302958e80f800
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
     2:        0x1025c421c - std::sys_common::backtrace::_print_fmt::h9819d35e2a5cda77
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/sys_common/backtrace.rs:68:5
     3:        0x1025c421c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h1f3776e0b5c7517d
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/sys_common/backtrace.rs:44:22
     4:        0x1025e3320 - core::fmt::rt::Argument::fmt::h626862aa6242248a
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/fmt/rt.rs:165:63
     5:        0x1025e3320 - core::fmt::write::heedef092c8c0962e
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/fmt/mod.rs:1157:21
     6:        0x1025c0958 - std::io::Write::write_fmt::h7178e8e2ea928914
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/io/mod.rs:1832:15
     7:        0x1025c4074 - std::sys_common::backtrace::_print::ha0f584bc7bfb9d2b
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/sys_common/backtrace.rs:47:5
     8:        0x1025c4074 - std::sys_common::backtrace::print::h417292deb95532ed
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/sys_common/backtrace.rs:34:9
     9:        0x1025c57a8 - std::panicking::default_hook::{{closure}}::h0cb68f1228c4613a
    10:        0x1025c549c - std::panicking::default_hook::h24535936bc1f51de
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/panicking.rs:298:9
    11:        0x1025c6060 - std::panicking::rust_panic_with_hook::h5db4d2345b297bed
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/panicking.rs:795:13
    12:        0x1025c5a90 - std::panicking::begin_panic_handler::{{closure}}::h3fd558f09a0d5492
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/panicking.rs:664:13
    13:        0x1025c46a4 - std::sys_common::backtrace::__rust_end_short_backtrace::hfc76eebe1ce501b2
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/sys_common/backtrace.rs:171:18
    14:        0x1025c5800 - rust_begin_unwind
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/panicking.rs:652:5
    15:        0x1026001fc - core::panicking::panic_fmt::hc2b459a5bd3dce66
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/panicking.rs:72:14
    16:        0x1026005e4 - core::result::unwrap_failed::h88d7eb352f20b747
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/result.rs:1654:5
    17:        0x102177e14 - core::result::Result<T,E>::expect::hd8965b36141f3d56
    18:        0x102181690 - build_script_build::generator::BindingGenerator::run::{{closure}}::hc5b069ba334404e7
    19:        0x1021909c4 - std::thread::scoped::scope::{{closure}}::h71fb584dc074dd60
    20:        0x102192fa0 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h495a652218759b2d
    21:        0x102190b24 - std::panicking::try::do_call::hd3fc0800c2662636
    22:        0x102192e98 - ___rust_try
    23:        0x102190ab8 - std::panicking::try::ha4ea956364f339b4
    24:        0x1021907e0 - std::thread::scoped::scope::hbf4af47715b70986
    25:        0x1021813a4 - build_script_build::generator::BindingGenerator::run::h9e287aacc7530955
    26:        0x1021809d4 - build_script_build::generator::BindingGenerator::generate_wrapper::h2df02edeb7e7dca1
    27:        0x1021a6564 - build_script_build::main::h92c244efe232b274
    28:        0x102193d1c - core::ops::function::FnOnce::call_once::hcb92fec8d7b86c39
    29:        0x10217d27c - std::sys_common::backtrace::__rust_begin_short_backtrace::h318d1c966e78bd08
    30:        0x10219052c - std::rt::lang_start::{{closure}}::h595fe049a0b92159
    31:        0x1025ba6ec - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h2f86a413382a979d
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/ops/function.rs:284:13
    32:        0x1025ba6ec - std::panicking::try::do_call::hd40c9eb4d233b111
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/panicking.rs:559:40
    33:        0x1025ba6ec - std::panicking::try::h13ac68ffa70c387b
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/panicking.rs:523:19
    34:        0x1025ba6ec - std::panic::catch_unwind::habea7b6fc986e966
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/panic.rs:149:14
    35:        0x1025ba6ec - std::rt::lang_start_internal::{{closure}}::h6b16436250c3cf62
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/rt.rs:141:48
    36:        0x1025ba6ec - std::panicking::try::do_call::h9970b928a0b20951
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/panicking.rs:559:40
    37:        0x1025ba6ec - std::panicking::try::h4dfbe3cb4cc8f253
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/panicking.rs:523:19
    38:        0x1025ba6ec - std::panic::catch_unwind::hf6a5e1e8ce5a10f5
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/panic.rs:149:14
    39:        0x1025ba6ec - std::rt::lang_start_internal::hecc68fef83c8f44d
                                 at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/rt.rs:141:20
    40:        0x102190504 - std::rt::lang_start::h7ecc524e215ec39d
    41:        0x1021a6cc4 - _main
warning: build failed, waiting for other jobs to finish...
```
This implies an issue with `open-cv` crate. Upon investigation, the following issue gave some clues: https://github.com/twistedfall/opencv-rust/issues/548, specifically https://github.com/twistedfall/opencv-rust/issues/548#issuecomment-2052615736 suggests that the bug is caused by https://github.com/twistedfall/opencv-rust/blob/985af46844b74581d38325561b5baa240292c797/binding-generator/src/field.rs#L153 with a PR to be merged found at https://github.com/KyleMayes/clang-rs/pull/58 that will fix things.

Later, it was announced in https://github.com/twistedfall/opencv-rust/issues/548#issuecomment-2093266462 that an update to `opencv-rust` should resolve this if bumped to `opencv = "=0.91.3"`.

Though, because of this update to `mat2image` crate: https://github.com/rcastill/mat2image/commit/7bc3cc72ecb814c524337b427fd25b20ce7f1803 it seems to cause the following error when fixing to the `opencv-rust` version above:
```
   Compiling pup v0.1.0 (/Users/tallam/github/tallamjr/origin/pup)
error[E0599]: the method `to_image` exists for struct `Mat`, but its trait bounds were not satisfied
     --> src/main.rs:452:40
      |
452   |             let original_image = frame.to_image()?;
      |                                        ^^^^^^^^ method cannot be called on `Mat` due to unsatisfied trait bounds
      |
     ::: /Users/tallam/github/tallamjr/origin/pup/target/debug/build/opencv-9aef8a2551c7da9a/out/opencv/core.rs:19719:2
      |
19719 |     pub struct Mat {
      |     -------------- doesn't satisfy `_: MatTraitConstManual` or `opencv::prelude::Mat: ToImage`
      |
      = note: the following trait bounds were not satisfied:
              `opencv::prelude::Mat: opencv::manual::core::mat::MatTraitConstManual`
              which is required by `opencv::prelude::Mat: ToImage`

warning: unused import: `mat2image::ToImage`
  --> src/main.rs:12:5
   |
12 | use mat2image::ToImage;
   |     ^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

For more information about this error, try `rustc --explain E0599`.
warning: `pup` (bin "pup") generated 1 warning
error: could not compile `pup` (bin "pup") due to 1 previous error; 1 warning emitted
```
Then if we fix `mat2image` crate to a revision immediately prior to that change, i.e. in the `Cargo.toml` set:
```
mat2image = { git = "https://github.com/rcastill/mat2image.git", rev = "f222cb14"}
```
we in theory should be Ok, but the following error still occurs:
```
   Compiling pup v0.1.0 (/Users/tallam/github/tallamjr/origin/pup)
error[E0599]: no method named `to_image` found for struct `opencv::prelude::Mat` in the current scope
   --> src/main.rs:452:40
    |
452 |             let original_image = frame.to_image()?;
    |                                        ^^^^^^^^ method not found in `Mat`

warning: unused import: `mat2image::ToImage`
  --> src/main.rs:12:5
   |
12 | use mat2image::ToImage;
   |     ^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

For more information about this error, try `rustc --explain E0599`.
warning: `pup` (bin "pup") generated 1 warning
error: could not compile `pup` (bin "pup") due to 1 previous error; 1 warning emitted

```
This can go one of two ways, either put on ice until tools are updated elsewhere, or try to use Github Actions to build for other platforms than macOS and compare outputs. TBD 🤔 ...

--- a/.cargo/config.toml
+++ b/.cargo/config.toml
@@ -13,5 +13,11 @@
 # [target.aarch64-apple-darwin]
 # rustflags = ["-C", "link-arg=-fuse-ld=/opt/homebrew/opt/llvm/bin/ld64.lld"]
 
+[target.aarch64-apple-darwin]
+rustflags = [
+  "-C", "link-arg=-undefined",
+  "-C", "link-arg=dynamic_lookup",
+]
+
 # [target.armv7-unknown-linux-gnueabihf]
 # linker = "arm-linux-gnueabihf-gcc"

--- a/.cargo/config.toml
+++ b/.cargo/config.toml
@@ -7,17 +7,19 @@
 # [target.x86_64-pc-windows-gnu]
 # rustflags = ["-C", "link-arg=-fuse-ld=lld"]
 
-# [target.x86_64-apple-darwin]
-# rustflags = ["-C", "link-arg=-fuse-ld=lld"]
+[target.x86_64-apple-darwin]
+rustflags = ["-C", "link-arg=-fuse-ld=lld"]
 
 # [target.aarch64-apple-darwin]
-# rustflags = ["-C", "link-arg=-fuse-ld=/opt/homebrew/opt/llvm/bin/ld64.lld"]
 
 [target.aarch64-apple-darwin]
 rustflags = [
   "-C", "link-arg=-undefined",
   "-C", "link-arg=dynamic_lookup",
+  # Xcode 16 beta linker.
+  "-C", "link-arg=-fuse-ld=/Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld",
+  "-Zlocation-detail=none", # `Z` is only accepted on the nightly compiler: `rustup default nightly`
 ]
 
-# [target.armv7-unknown-linux-gnueabihf]
-# linker = "arm-linux-gnueabihf-gcc"
+[target.armv7-unknown-linux-gnueabihf]
+linker = "arm-linux-gnueabihf-gcc"

--- a/.config/brew/Brewfile
+++ b/.config/brew/Brewfile
@@ -1,0 +1,1 @@
+brew 'opencv'

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -1,0 +1,51 @@
+name: build
+
+on: [ push, pull_request ]
+
+jobs:
+  default:
+    name: Default
+    strategy:
+      matrix:
+        platform: [ ubuntu-latest, macos-latest ]
+        toolchain: [ stable, nightly ]
+    runs-on: ${{ matrix.platform }}
+    steps:
+      - uses: actions/checkout@v4
+      - uses: dtolnay/rust-toolchain@master
+        with:
+          toolchain: ${{ matrix.toolchain }}
+      - name: Build
+        run: cargo build --release --all-features
+  no_std:
+    name: no_std
+    strategy:
+      matrix:
+        platform: [ ubuntu-latest, macos-latest ]
+        toolchain: [ stable, nightly ]
+    runs-on: ${{ matrix.platform }}
+    steps:
+      - uses: actions/checkout@v4
+      - uses: dtolnay/rust-toolchain@master
+        with:
+          toolchain: ${{ matrix.toolchain }}
+      - name: Build
+        working-directory: no_std/nix
+        run: cargo build --release --all-features
+  build_std:
+    name: build-std
+    strategy:
+      matrix:
+        project_dir: [ build_std, no_main/nix ]
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: dtolnay/rust-toolchain@master
+        with:
+          toolchain: nightly
+      - name: Build
+        working-directory: ${{ matrix.project_dir }}
+        run: >
+          rustup component add rust-src;
+          RUSTFLAGS="-Zlocation-detail=none" cargo +nightly build -Z build-std=std,panic_abort -Z build-std-features="optimize_for_size" --target x86_64-unknown-linux-gnu --release;
+          cargo +nightly build -Z build-std=std,panic_abort -Z build-std-features=panic_immediate_abort --target x86_64-unknown-linux-gnu --release;

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -60,6 +60,7 @@ jobs:
         run: |
           env HOMEBREW_NO_AUTO_UPDATE=1 brew install miniforge
           env HOMEBREW_NO_AUTO_UPDATE=1 brew install opencv || true
+          export DYLD_FALLBACK_LIBRARY_PATH="$(xcode-select --print-path)/usr/lib/"
         if: matrix.platform.os_name == 'macOS-aarch64'
       # Compile, build.
       - name: Build binaries

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -17,26 +17,26 @@ jobs:
       fail-fast: false
       matrix:
         platform:
-          - os_name: Linux-aarch64
-            os: ubuntu-20.04
-            target: aarch64-unknown-linux-musl
-            bin: figctl-linux-arm64
-          - os_name: Linux-x86_64
-            os: ubuntu-20.04
-            target: x86_64-unknown-linux-gnu
-            bin: figctl-linux-amd64
-          - os_name: Windows-x86_64
-            os: windows-latest
-            target: x86_64-pc-windows-msvc
-            bin: figctl-amd64.exe
-          - os_name: macOS-x86_64
-            os: macOS-latest
-            target: x86_64-apple-darwin
-            bin: figctl-darwin-amd64
+          # - os_name: Linux-aarch64
+          #   os: ubuntu-20.04
+          #   target: aarch64-unknown-linux-musl
+          #   bin: pup-linux-arm64
+          # - os_name: Linux-x86_64
+          #   os: ubuntu-20.04
+          #   target: x86_64-unknown-linux-gnu
+          #   bin: pup-linux-amd64
+          # - os_name: Windows-x86_64
+          #   os: windows-latest
+          #   target: x86_64-pc-windows-msvc
+          #   bin: pup-amd64.exe
+          # - os_name: macOS-x86_64
+          #   os: macOS-latest
+          #   target: x86_64-apple-darwin
+          #   bin: pup-darwin-amd64
           - os_name: macOS-aarch64
-            os: macOS-latest
+            os: macOS-13
             target: aarch64-apple-darwin
-            bin: figctl-darwin-arm64
+            bin: pup-darwin-arm64
         toolchain:
           - stable
     steps:
@@ -50,10 +50,10 @@ jobs:
           args: "--release"
           strip: true
       - name: Rename binary (linux and macos)
-        run: mv target/${{ matrix.platform.target }}/release/figctl target/${{ matrix.platform.target }}/release/${{ matrix.platform.bin }}
+        run: mv target/${{ matrix.platform.target }}/release/pup target/${{ matrix.platform.target }}/release/${{ matrix.platform.bin }}
         if: matrix.platform.os_name != 'Windows-x86_64'
       - name: Rename binary (windows)
-        run: mv target/${{ matrix.platform.target }}/release/figctl.exe target/${{ matrix.platform.target }}/release/${{ matrix.platform.bin }}
+        run: mv target/${{ matrix.platform.target }}/release/pup.exe target/${{ matrix.platform.target }}/release/${{ matrix.platform.bin }}
         if: matrix.platform.os_name == 'Windows-x86_64'
       - name: Generate SHA-256
         run: shasum -a 256 target/${{ matrix.platform.target }}/release/${{ matrix.platform.bin }} | cut -d ' ' -f 1 > target/${{ matrix.platform.target }}/release/${{ matrix.platform.bin }}.sha256

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -41,7 +41,26 @@ jobs:
           - stable
     steps:
       - uses: actions/checkout@v3
-      - name: Build binary
+      # Install opencv on macOS
+      - name: Update Homebrew
+        run: |
+          brew update --preinstall
+          cat "$(brew --repository)/Library/Taps/homebrew/homebrew-core/Formula/foo.rb" > .github/brew-formulae
+        if: matrix.platform.os_name == 'macOS-aarch64'
+      - name: Configure Homebrew cache
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/Library/Caches/Homebrew/foo--*
+            ~/Library/Caches/Homebrew/downloads/*--foo-*
+          key: brew-${{ hashFiles('.github/brew-formulae') }}
+          restore-keys: brew-
+      - name: Install Homebrew dependencies
+        run: |
+          env HOMEBREW_NO_AUTO_UPDATE=1 brew install opencv
+        if: matrix.platform.os_name == 'macOS-aarch64'
+      # Compile, build.
+      - name: Build binaries
         uses: houseabsolute/actions-rust-cross@v0
         with:
           command: "build"

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -45,19 +45,24 @@ jobs:
       - name: Update Homebrew
         run: |
           brew update --preinstall
-          cat "$(brew --repository)/Library/Taps/homebrew/homebrew-core/Formula/foo.rb" > .github/brew-formulae
         if: matrix.platform.os_name == 'macOS-aarch64'
       - name: Configure Homebrew cache
         uses: actions/cache@v2
         with:
           path: |
-            ~/Library/Caches/Homebrew/foo--*
-            ~/Library/Caches/Homebrew/downloads/*--foo-*
+            ~/Library/Caches/Homebrew/opencv*
+            ~/Library/Caches/Homebrew/miniforge*
+            ~/Library/Caches/Homebrew/downloads/*--opencv-*
+            ~/Library/Caches/Homebrew/downloads/*--miniforge-*
           key: brew-${{ hashFiles('.github/brew-formulae') }}
           restore-keys: brew-
       - name: Install Homebrew dependencies
         run: |
+          rm '/usr/local/bin/2to3'
+          env HOMEBREW_NO_AUTO_UPDATE=1 brew install miniforge
+          brew unlink python@3.12
           env HOMEBREW_NO_AUTO_UPDATE=1 brew install opencv
+          brew link --overwrite python@3.12
         if: matrix.platform.os_name == 'macOS-aarch64'
       # Compile, build.
       - name: Build binaries

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -58,11 +58,8 @@ jobs:
           restore-keys: brew-
       - name: Install Homebrew dependencies
         run: |
-          rm '/usr/local/bin/2to3'
           env HOMEBREW_NO_AUTO_UPDATE=1 brew install miniforge
-          brew unlink python@3.12
-          env HOMEBREW_NO_AUTO_UPDATE=1 brew install opencv
-          brew link --overwrite python@3.12
+          env HOMEBREW_NO_AUTO_UPDATE=1 brew install opencv || true
         if: matrix.platform.os_name == 'macOS-aarch64'
       # Compile, build.
       - name: Build binaries

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -45,24 +45,19 @@ jobs:
       - name: Update Homebrew
         run: |
           brew update --preinstall
+          brew bundle --file .config/brew/Brewfile || true
         if: matrix.platform.os_name == 'macOS-aarch64'
-      - name: Configure Homebrew cache
-        uses: actions/cache@v2
+      - name: Setup Xcode
+        uses: maxim-lobanov/setup-xcode@v1
         with:
-          path: |
-            ~/Library/Caches/Homebrew/opencv*
-            ~/Library/Caches/Homebrew/miniforge*
-            ~/Library/Caches/Homebrew/downloads/*--opencv-*
-            ~/Library/Caches/Homebrew/downloads/*--miniforge-*
-          key: brew-${{ hashFiles('.github/brew-formulae') }}
-          restore-keys: brew-
-      - name: Install Homebrew dependencies
+          xcode-version: '14.3.1'
+      - name: Install LLVM and Clang
+        uses: KyleMayes/install-llvm-action@v2
+        with:
+          version: "10.0"
+      - name: OpenCV Troubleshooting
         run: |
-          env HOMEBREW_NO_AUTO_UPDATE=1 brew install miniforge
-          env HOMEBREW_NO_AUTO_UPDATE=1 brew install opencv || true
           export DYLD_FALLBACK_LIBRARY_PATH="$(xcode-select --print-path)/usr/lib/"
-        if: matrix.platform.os_name == 'macOS-aarch64'
-      # Compile, build.
       - name: Build binaries
         uses: houseabsolute/actions-rust-cross@v0
         with:

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -2,7 +2,7 @@ name: build and release
 
 on:
   push:
-    branches: [ "master" ]
+    branches: [ "issues/1/m2build" ]
   pull_request:
     branches: [ "master" ]
 

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -3,14 +3,12 @@ name = "pup"
 version = "0.1.0"
 edition = "2021"
 
-# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
-
 [dependencies]
 anyhow = "1.0.79"
-candle = { git = "https://github.com/huggingface/candle.git", version = "0.4.0" , package = "candle-core"}
-candle-nn = { git = "https://github.com/huggingface/candle.git", version = "0.4.0" }
-candle-onnx = { git = "https://github.com/huggingface/candle.git", version = "0.4.0" }
-candle-transformers = { git = "https://github.com/huggingface/candle.git", version = "0.4.0" }
+candle = { git = "https://github.com/huggingface/candle.git", version = "0.6.0", features = ["metal"], package = "candle-core"}
+candle-nn = { git = "https://github.com/huggingface/candle.git", version = "0.6.0" }
+candle-onnx = { git = "https://github.com/huggingface/candle.git", version = "0.6.0" }
+candle-transformers = { git = "https://github.com/huggingface/candle.git", version = "0.6.0" }
 clap = { features = ["derive"], version = "4.4.18" }
 image = { version = "0.24.7", default-features = false, features = ["jpeg", "png"] }
 imageproc = { version = "0.23.0", default-features = false }
@@ -21,6 +19,7 @@ tracing = "0.1.37"
 tracing-chrome = "0.7.1"
 tracing-subscriber = "0.3.7"
 rusttype = "0.9.3"
-opencv = "0.88.8"
-mat2image = { git = "https://github.com/rcastill/mat2image.git" }
+mat2image = { git = "https://github.com/rcastill/mat2image.git", rev = "f222cb14"}
+# mat2image = { git = "https://github.com/rcastill/mat2image.git"}
 openssl = { version = "0.10", features = ["vendored"] }
+opencv = "=0.91.3"

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -3,6 +3,7 @@ name = "pup"
 version = "0.1.0"
 edition = "2021"
 
+[target.'cfg(target_vendor= "apple")'.dependencies]
 [dependencies]
 anyhow = "1.0.79"
 candle = { git = "https://github.com/huggingface/candle.git", version = "0.6.0", features = ["metal"], package = "candle-core"}
@@ -22,4 +23,4 @@ rusttype = "0.9.3"
 mat2image = { git = "https://github.com/rcastill/mat2image.git", rev = "f222cb14"}
 # mat2image = { git = "https://github.com/rcastill/mat2image.git"}
 openssl = { version = "0.10", features = ["vendored"] }
-opencv = "=0.91.3"
+opencv = "=0.88.8"

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,26 +1,31 @@
 [package]
+edition = "2021"
 name = "pup"
 version = "0.1.0"
-edition = "2021"
 
 [target.'cfg(target_vendor= "apple")'.dependencies]
 [dependencies]
 anyhow = "1.0.79"
 candle = { git = "https://github.com/huggingface/candle.git", version = "0.6.0", features = ["metal"], package = "candle-core"}
-candle-nn = { git = "https://github.com/huggingface/candle.git", version = "0.6.0" }
-candle-onnx = { git = "https://github.com/huggingface/candle.git", version = "0.6.0" }
-candle-transformers = { git = "https://github.com/huggingface/candle.git", version = "0.6.0" }
+candle-nn = { git = "https://github.com/huggingface/candle.git", version = "0.6.0", features = ["metal"]}
+candle-onnx = { git = "https://github.com/huggingface/candle.git", version = "0.6.0"}
+candle-transformers = { git = "https://github.com/huggingface/candle.git", version = "0.6.0", features = ["metal"]}
 clap = { features = ["derive"], version = "4.4.18" }
+hf-hub = "0.3.2"
 image = { version = "0.24.7", default-features = false, features = ["jpeg", "png"] }
 imageproc = { version = "0.23.0", default-features = false }
-hf-hub = "0.3.2"
+opencv = "0.92.0"
+openssl = { version = "0.10", features = ["vendored"] }
+rusttype = "0.9.3"
 serde = "1.0.196"
 serde_json = "1.0.113"
 tracing = "0.1.37"
 tracing-chrome = "0.7.1"
 tracing-subscriber = "0.3.7"
-rusttype = "0.9.3"
-mat2image = { git = "https://github.com/rcastill/mat2image.git", rev = "f222cb14"}
-# mat2image = { git = "https://github.com/rcastill/mat2image.git"}
-openssl = { version = "0.10", features = ["vendored"] }
-opencv = "=0.88.8"
+
+[profile.release]
+codegen-units = 1   # Reduce number of codegen units to increase optimizations.
+lto = true          # Enable Link Time Optimization
+opt-level = "z"     # Optimize for size.
+panic = "abort"     # Abort on panic
+strip = true        # Automatically strip symbols from the binary.

--- a/__build.rs
+++ b/__build.rs
@@ -1,0 +1,52 @@
+use std::env;
+use std::process::Command;
+
+fn main() {
+    // Run the sed command
+    let output = Command::new("sh")
+        .arg("-c")
+        .arg("rust -vV | sed -n 's/^host: \\(.*\\)$/\\1/p'")
+        .output()
+        .expect("Failed to execute command");
+
+    // Check if the command was successful
+    if output.status.success() {
+        let result = std::str::from_utf8(&output.stdout).expect("Failed to parse output");
+        println!("Extracted host: {}", result.trim());
+    } else {
+        let error = std::str::from_utf8(&output.stderr).expect("Failed to parse error");
+        eprintln!("Error: {}", error);
+    }
+
+    // Set an environment variable based on the extracted host
+    println!("cargo:rustc-env=HOST_ARCH={}", result.trim());
+
+    // Retrieve the target triple
+    let target = env::var("HOST_ARCH").unwrap();
+
+    let xcode_path = Command::new("sh")
+        .arg("-c")
+        .arg("xcode-select -p")
+        .output()
+        .expect("Failed to execute command");
+
+    if output.status.success() {
+        let path = std::str::from_utf8(&output.stdout).expect("Failed to parse output");
+        println!("Extracted host: {}", result.trim());
+    } else {
+        let error = std::str::from_utf8(&output.stderr).expect("Failed to parse error");
+        eprintln!("Error: {}", error);
+    }
+
+    // Check if the target is aarch64-apple-darwin
+    if target == "aarch64-apple-darwin" {
+        // Set custom RUSTFLAGS for the target
+        let rustflags = format!("-C link-arg=-fuse-ld={}/usr/bin/ld", path.trim());
+
+        // Export the RUSTFLAGS environment variable
+        println!("cargo:rustc-env=RUSTFLAGS={}", rustflags);
+    }
+
+    // Optionally, run any other build script actions here
+    println!("cargo:rerun-if-changed=build.rs");
+}

--- a/src/main.rs
+++ b/src/main.rs
@@ -4,12 +4,9 @@ extern crate intel_mkl_src;
 #[cfg(feature = "accelerate")]
 extern crate accelerate_src;
 
-use image::DynamicImage;
+use image::{DynamicImage, ImageBuffer};
 
 use opencv::{core, highgui, prelude::*, videoio};
-
-// use cv_convert::{FromCv, IntoCv, TryFromCv, TryIntoCv};
-use mat2image::ToImage;
 
 mod model;
 use model::{Multiples, YoloV8, YoloV8Pose};
@@ -391,6 +388,22 @@ fn dynamic_image_to_mat(image: DynamicImage) -> opencv::Result<Mat> {
     Ok(mat)
 }
 
+fn mat_to_dynamic_image(mat: &Mat) -> DynamicImage {
+    let mat_size = mat.size().unwrap();
+    let width = mat_size.width;
+    let height = mat_size.height;
+    // let mat_type = mat.typ();
+
+    // let channels = core::CV_8UC3 as u8;
+    // assert!(channels == 3, "Expected 3 channels for RGB image");
+
+    let data = mat.data_bytes().unwrap();
+    let image_buffer: ImageBuffer<image::Rgb<u8>, Vec<u8>> =
+        ImageBuffer::from_raw(width as u32, height as u32, data.to_vec()).unwrap();
+
+    DynamicImage::ImageRgb8(image_buffer)
+}
+
 pub fn run<T: Task>(args: Args) -> anyhow::Result<()> {
     let device = pup::device(args.cpu)?;
     // Create the model and load the weights from the file.
@@ -449,7 +462,8 @@ pub fn run<T: Task>(args: Args) -> anyhow::Result<()> {
             // Process the nth frame
             println!("Processing frame {}", frame_count);
 
-            let original_image = frame.to_image()?;
+            // let original_image = frame.to_image()?;
+            let original_image = mat_to_dynamic_image(&frame);
 
             let (width, height) = {
                 let w = original_image.width() as usize;
@@ -504,7 +518,8 @@ pub fn run<T: Task>(args: Args) -> anyhow::Result<()> {
                 0,
             )?;
 
-            highgui::imshow(window, &bgr_img_mat)?;
+            // highgui::imshow(window, &bgr_img_mat)?;
+            highgui::imshow(window, &frame)?;
 
             let key = highgui::wait_key(1)?;
             if key > 0 && key != 255 {

